File: mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingRowViewModelsBuilder.java
package com.alexstyl.specialdates.upcoming;

import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.date.DateComparator;
import com.alexstyl.specialdates.date.TimePeriod;
import com.alexstyl.specialdates.events.bankholidays.BankHoliday;
import com.alexstyl.specialdates.events.namedays.NamesInADate;
import com.alexstyl.specialdates.util.HashMapList;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

final class UpcomingRowViewModelsBuilder {

    private static final List<UpcomingRowViewModel> NO_CELEBRATIONS = Collections.emptyList();
    private static final List<ContactEvent> NO_CONTACT_EVENTS = Collections.emptyList();
    private final DateComparator dateComparator = DateComparator.INSTANCE;

    private final HashMapList<AnnualDate, ContactEvent> contactEvents = new HashMapList<>();
    private final HashMap<AnnualDate, NamesInADate> namedays = new HashMap<>();
    private final HashMap<AnnualDate, BankHoliday> bankHolidays = new HashMap<>();
    private final TimePeriod duration;
    private final UpcomingEventRowViewModelFactory upcomingRowViewModelFactory;
    private final UpcomingEventsAdRules adRules;

    UpcomingRowViewModelsBuilder(TimePeriod duration, UpcomingEventRowViewModelFactory upcomingRowViewModelFactory, UpcomingEventsAdRules adRules) {
        this.duration = duration;
        this.upcomingRowViewModelFactory = upcomingRowViewModelFactory;
        this.adRules = adRules;
    }

    UpcomingRowViewModelsBuilder withContactEvents(List<ContactEvent> contactEvents) {
        for (ContactEvent contactEvent : contactEvents) {
            Date date = contactEvent.getDate();
            this.contactEvents.addValue(new AnnualDate(date), contactEvent);
        }
        return this;
    }

    UpcomingRowViewModelsBuilder withNamedays(List<NamesInADate> namedays) {
        for (NamesInADate nameday : namedays) {
            Date date = nameday.getDate();
            this.namedays.put(new AnnualDate(date), nameday);
        }
        return this;
    }

    UpcomingRowViewModelsBuilder withBankHolidays(List<BankHoliday> bankHolidays) {
        for (BankHoliday bankHoliday : bankHolidays) {
            Date date = bankHoliday.getDate();
            this.bankHolidays.put(new AnnualDate(date), bankHoliday);
        }
        return this;
    }

    public List<UpcomingRowViewModel> build() {
        if (noEventsArePresent()) {
            return NO_CELEBRATIONS;
        }

        List<UpcomingRowViewModel> rowsViewModels = new ArrayList<>();
        Date indexDate = duration.getStartingDate();
        Date lastDate = duration.getEndingDate();

        Date previousDate = Date.startOfTheYear(indexDate.getYear());

        int index = 0;
        while (dateComparator.compare(indexDate, lastDate) <= 0) {
            AnnualDate annualDate = new AnnualDate(indexDate);
            if (containsAnyEventsOn(annualDate)) {
                if (indexDate.getYear() != previousDate.getYear()) {
                    rowsViewModels.add(upcomingRowViewModelFactory.createRowForYear(indexDate.getYear()));
                }
                if (indexDate.getMonth() != previousDate.getMonth()) {
                    if (adRules.shouldAppendAdForEndOfMonth(index)) {
                        rowsViewModels.add(new AdViewModel(indexDate));
                    }
                    rowsViewModels.add(upcomingRowViewModelFactory.createRowForMonth(indexDate.getMonth()));
                    adRules.onNewMonthAdded(index);
                }
                if (adRules.shouldAppendAdAt(index)) {
                    rowsViewModels.add(new AdViewModel(indexDate));
                    adRules.onNewAdAdded(index);
                }
                UpcomingRowViewModel viewModel = upcomingRowViewModelFactory.createEventViewModel(
                        indexDate,
                        getEventsOn(annualDate),
                        this.namedays.get(annualDate),
                        bankHolidays.get(annualDate)
                );
                rowsViewModels.add(viewModel);
                previousDate = indexDate;
                index++;
            }
            indexDate = indexDate.addDay(1);
        }
        return rowsViewModels;
    }

    private List<ContactEvent> getEventsOn(AnnualDate indexDate) {
        List<ContactEvent> contactEvent = contactEvents.get(indexDate);
        if (contactEvent == null) {
            return NO_CONTACT_EVENTS;
        } else {
            return contactEvent;
        }
    }

    private boolean noEventsArePresent() {
        return contactEvents.isEmpty() && namedays.isEmpty() && bankHolidays.isEmpty();
    }

    private boolean containsAnyEventsOn(AnnualDate date) {
        return getEventsOn(date).size() > 0 || namedays.containsKey(date) || bankHolidays.containsKey(date);
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingRowViewModelsBuilder.kt
package com.alexstyl.specialdates.upcoming

import com.alexstyl.specialdates.date.ContactEvent
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.date.DateComparator
import com.alexstyl.specialdates.date.TimePeriod
import com.alexstyl.specialdates.events.bankholidays.BankHoliday
import com.alexstyl.specialdates.events.namedays.NamesInADate
import com.alexstyl.specialdates.util.HashMapList

class UpcomingRowViewModelsBuilder(private val duration: TimePeriod,
                                   private val viewModelFactory: UpcomingEventRowViewModelFactory,
                                   private val adRules: UpcomingEventsAdRules
) {

    private val dateComparator = DateComparator.INSTANCE
    private val contactEvents = HashMapList<Date, ContactEvent>()
    private val namedays = HashMap<Date, NamesInADate>()
    private val bankHolidays = HashMap<Date, BankHoliday>()

    fun withContactEvents(contactEvents: List<ContactEvent>): UpcomingRowViewModelsBuilder {
        this.contactEvents.clear()
        for (contactEvent in contactEvents) {
            val annualDate = contactEvent.date
            this.contactEvents.addValue(annualDate, contactEvent)
        }
        return this
    }

    fun withNamedays(namedays: List<NamesInADate>): UpcomingRowViewModelsBuilder {
        this.namedays.clear()
        for (nameday in namedays) {
            val date = nameday.date
            this.namedays.put(date, nameday)
        }
        return this
    }

    fun withBankHolidays(bankHolidays: List<BankHoliday>): UpcomingRowViewModelsBuilder {
        this.bankHolidays.clear()
        for (bankHoliday in bankHolidays) {
            val date = bankHoliday.date
            this.bankHolidays.put(date, bankHoliday)
        }
        return this
    }

    fun build(): List<UpcomingRowViewModel> {
        if (noEventsArePresent()) {
            return NO_CELEBRATIONS
        }

        val rowsViewModels = ArrayList<UpcomingRowViewModel>()
        var indexDate = duration.startingDate
        val lastDate = duration.endingDate

        var index = 0
        while (dateComparator.compare(indexDate, lastDate) <= 0) {
            val annualDate = indexDate
            if (containsAnyEventsOn(annualDate)) {

                rowsViewModels.add(viewModelFactory.createDateHeader(indexDate))

                if (bankHolidays[annualDate] != null) {
                    rowsViewModels.add(viewModelFactory.createViewModelFor(bankHolidays[annualDate]!!))
                }
                if (namedays[annualDate] != null) {
                    rowsViewModels.add(viewModelFactory.createViewModelFor(namedays[annualDate]!!))
                }

                getEventsOn(annualDate).forEach {
                    rowsViewModels.add(viewModelFactory.createViewModelFor(it))

                }
                if (adRules.shouldAppendAd()) {
                    rowsViewModels.add(AdViewModel(indexDate))
                    adRules.onNewAdAdded()
                }
                index++
            }
            indexDate = indexDate.addDay(1)
        }
        return rowsViewModels
    }

    private fun getEventsOn(indexDate: Date): List<ContactEvent> {
        val contactEvent = contactEvents.get(indexDate)
        if (contactEvent == null) {
            return NO_CONTACT_EVENTS
        } else {
            return contactEvent
        }
    }

    private fun noEventsArePresent(): Boolean {
        return contactEvents.isEmpty && namedays.isEmpty() && bankHolidays.isEmpty()
    }

    private fun containsAnyEventsOn(date: Date): Boolean {
        return getEventsOn(date).size > 0 || namedays.containsKey(date) || bankHolidays.containsKey(date)
    }

    companion object {

        private val NO_CELEBRATIONS = emptyList<UpcomingRowViewModel>()
        private val NO_CONTACT_EVENTS = emptyList<ContactEvent>()
    }
}
